Slider Handle Widget

thomascft opened the pull request:
I've added a `handle_widget` to `wibox.widget.slider`. I'm not sure how to do the tests, but I'd be willing to try writing them if I was given a direction. I'm not sure on the docs, so if I could get more details on the formatting of those that'd be great!

![Slider Widget](https://i.imgur.com/xQhSFUt.png)

Elv13 commented on lib/wibox/widget/slider.lua:
```diff
@@ -318,6 +319,7 @@ function slider:set_value(value)
     if changed then
         self:emit_signal( "property::value", value)
         self:emit_signal( "widget::redraw_needed" )
+        self:emit_signal("widget::layout_changed")
```
Please add an `if self._private.handle_widget then` around this.

You also want here is to create a custom setter for `set_handle_widget` instead of using the standard ones in the loop at line 292 so it can have the signal too (copy/paste the boilerplate code).

thomascft replied:
Done I think

Elv13 commented on lib/wibox/widget/slider.lua:
```diff
@@ -271,6 +271,7 @@ local properties = {
     handle_width         = false,
     handle_border_width  = 0,
     handle_border_color  = false,
+    handle_widget       = false,
```
Please add a `@property` section for this. You can copy/paste one and modify it. Also make sure to copy paste the corresponding file in `tests/examples/wibox/wiget/progressbar` and modify it to demonstrate how this property works. You need `ldoc` installed for this to work. You can open the doc in your local browser to see how it renders. No server required, they are static HTML files.

thomascft replied:
I assume you meant `tests/examples/wibox/widget/slider`. I was looking through those while I was trying to get the documentation working. I couldn't get the properties to show up in the `Object Properties` header, and I assume it has something to do with this. What file exactly should I edit?

Elv13 commented on lib/wibox/widget/slider.lua:
```diff
@@ -505,6 +507,45 @@ function slider:fit(_, width, height)
     return width, height
 end
 
+function slider:layout(context, width, height)
```
Maybe add `if not self._private.handle_widget then return {} end`? That depends if you add / plan_to_add a bar widget too.

Elv13 commented on lib/wibox/widget/slider.lua:
```diff
@@ -505,6 +507,45 @@ function slider:fit(_, width, height)
+        local x_offset, y_offset = 0, 0
+
+        if margins then
+            if type(margins) == "number" then
```
Please create a custom setter for `set_handle_margins` to always convert the margins to a table at the setter level. This is a mini API change, so please add a note in the `behavior changes` section of `doc/89-NEWS.md`.

You can then remove the type code from `:draw()`. This way we wont have duplicated type checking code. You will also have less tests to write since that `if` will be gone.

like
```lua
function slider:set_handle_margins(value)
    if type(value) == "number" then
        value = {
            top    = value,
            bottom = value,
            left   = value,
            right  = value,
         }
    end

    self._private.handle_margins = value
    self:emit_signal( "property::handle_margins", value)
    self:emit_signal( "widget::redraw_needed" )
end
```

Elv13 reviewed (changes requested):
Thanks for this. It looks like an useful feature. I left a few minor comments.

For the tests, if you want, you can create some unit tests in `spec/` and run them using `make check-unit`. You need `busted` to be installed. We don't have any for the slider right now. See `spec/wibox/widget/slider_spec.lua` for a minimal example.

**But** an alternative is to create a couple documentation examples based tests. They are not as fancy, but a lot easier to do and they provide much needed code examples for the users. You can copy paste an existing one and modify it. Pay attention to `--DOC_HIDE` (`_START`/`_END`) and `__DOC_NEWLINE`. They are preprocessed to help format the examples to remove the boilerplate. We are trying to have at least one of those for every properties.

The automatic bar for the tests right now is 91% overage. This should be rather easy here. If the margin type `if` gets mutualized in the setter, then a simple test should hit nearly 100% of the code.

Bonus points if you can implement `bar_widget` for the slider (and maybe progressbar too?). The code would probably be 95% identical and would help consistency across those modules.

Elv13 commented:
Oh, and also please install `luacheck` and run it, there's a few linting errors:

```txt

Checking lib/wibox/widget/slider.lua              4 warnings

    lib/wibox/widget/slider.lua:517:15: value assigned to variable 'x' is overwritten on line 543 before use
    lib/wibox/widget/slider.lua:517:18: value assigned to variable 'y' is overwritten on line 544 before use
    lib/wibox/widget/slider.lua:531:17: value assigned to variable 'handle_height' is unused
    lib/wibox/widget/slider.lua:536:17: value assigned to variable 'handle_height' is unused

Checking lib/wibox/widget/systray.lua             OK
```

thomascft commented:
I'll clean up my code and add the docs. I was considering doing a rewrite of the slider widget at some point rather than trying to edit what exists. I think the current widget is more focused on cario drawings, and a rewrite/refresh might be better. I'm open to anything though.

Elv13 commented:
> I was considering doing a rewrite of the slider widget at some point rather than trying to edit what exists.

It would be pretty hard to merge (API compatibility and duplication requirements). IMHO extending the current one benefits more users. What kind of feature would such a rewrite improve upon/add? Not saying the current widget is incredible. It has a strange history.

> I think the current widget is more focused on cario drawing

I think it's useful to provide both shape/cairo and sub-widgets in the same module. Most users are satisfied with shapes+colors, but might eventually want to upgrade to a declarative custom sub-widgets. Doing so without having to replace the code or adding a 3rd party widget library makes it a lot more straightforward. Having less overlapping modules also reduce the mental load when learning/experimenting. It doesn't make the internal implementation pretty, but I am usually fine with internal complexity if it makes the user experience better.

thomascft commented:
> 

Understandable. For the bar widget, ow exactly would you like that to be implemented functionality-wise?  Would you like it to span the container, or only fill the active portion?

thomascft commented:
> It would be pretty hard to merge (API compatibility and duplication requirements). IMHO extending the current one benefits more users. What kind of feature would such a rewrite improve upon/add? Not saying the current widget is incredible. It has a strange history.

I just find the code to be a bit messy, and a refactor would make it easier to understand/maintain. I figure a refactor is vastly different from a complete rewrite, I just think the chance to clean it up a bit at some point might be valuable.

Elv13 commented:
> I just find the code to be a bit messy, and a refactor would make it easier to understand/maintain. I figure a refactor is vastly different from a complete rewrite, I just think the chance to clean it up a bit at some point might be valuable.

Refactors are fine. Just make sure the public API and behavior don't have unexpected change. Depending on what kind of scope you want in the PR, maybe you want the initial version merged first? That still requires the doc and tests/example/wibox/widget/slider file.

> Understandable. For the bar widget, how exactly would you like that to be implemented functionality-wise? Would you like it to span the container, or only fill the active portion?

Behave the same as the `bar_shape`, but with a widget instead of a shape?

thomascft commented:
I'll work on bar widget and the docs/tests. The refactor would definitely be a different pr.

thomascft commented:
I'm currently working on the margin setters. How would I call them on the widget init? I set them up to format the `self._private.*_margins` if a value isn't passed. I'd like to format so I ca always assume it's a table in the draw + layout functions.

Elv13 commented:
> How would I call them on the widget init?

Move `gtable.crush(ret._private, args or {})` 2 lines down and remove the `._private`. This will force all properties to be created using the setters.

This is already used in various other widgets where setters have some business logic.

thomascft commented:
So how does the setter get called?

Elv13 commented:
> So how does the setter get called?

AwesomeWM objects (both `gears.object` and `capi.object`) have meta code to make sure `obj:set_foo("bar")`,  `obj.foo = "bar"` and `obj["foo"] = "bar"` are equivalent. `gears.table.crush` for a for loop that sets all values from one table to the other table. So it will indirectly call the setters automatically.

thomascft commented:
Ok, I added the `bar_widget` and tried to make the margin setter functions. Still a bit confused on documentation. I added what I understood, but I don't think I covered it all. Are the code examples supposed to be in the widget file?

Aire-One commented on lib/wibox/widget/slider.lua:
```diff
@@ -318,6 +320,11 @@ function slider:set_value(value)
     if changed then
         self:emit_signal( "property::value", value)
         self:emit_signal( "widget::redraw_needed" )
+        if self._private.handle_widget then
+            self:emit_signal("widget::layout_changed")
+        elseif self._private.bar_widget then
+            self:emit_signal("widget::layout_changed")
+        end
```
Both branches have the same body. Please make them be part of the same if statement.

thomascft replied:
That was a desperate attempt at testing, Just fixed that.

Aire-One commented on lib/wibox/widget/slider.lua:
```diff
@@ -368,17 +409,11 @@ function slider:draw(_, cr, width, height)
+        bar_height = bar_height or (
+            height - (margins.top or 0) - (margins.bottom or 0)
+        )
+        x_offset, y_offset = margins.left or 0, margins.top or 0
```
Split these assignations on two different statements. The additional `or` clauses make the line hard to read.

thomascft replied:
Like this?
```lua
        x_offset, y_offset = margins.left or 0,
            margins.top or 0
```

Aire-One commented on lib/wibox/widget/slider.lua:
```diff
@@ -458,17 +493,11 @@ function slider:draw(_, cr, width, height)
-            handle_height = handle_height -
-                (margins.top  or 0) - (margins.bottom or 0)
-        end
+        x_offset, y_offset = margins.left or 0, margins.top or 0
```
Split this line

Aire-One commented on lib/wibox/widget/slider.lua:
```diff
@@ -458,17 +493,11 @@ function slider:draw(_, cr, width, height)
-        end
+        x_offset, y_offset = margins.left or 0, margins.top or 0
+        handle_width  = handle_width  -
+            (margins.left or 0) - (margins.right  or 0)
+        handle_height = handle_height -
+            (margins.top  or 0) - (margins.bottom or 0)
```
If we are going to use X times the `margins.direction or 0` statement, we should definitely have local variables to handle these operations once for all.

thomascft replied:
I'm not sure if I fully understand. What scope should the vars be in?

thomascft replied:
Ok, just pushed some changes to the margin stuff. If you could take a look and see if it's what you were looking for.

Aire-One commented on lib/wibox/widget/slider.lua:
```diff
@@ -505,6 +534,68 @@ function slider:fit(_, width, height)
+    local bar_height, bar_width = self._private.bar_height or height, width
+    local handle_height, handle_width = height, self._private.handle_width
+        or beautiful.slider_handle_width
+        or math.floor(height / 2)
```
Is there a particular reason to make the width defaults to half the height?

I would argue we can make it a square and save this `math.floor` + division.

thomascft replied:
I believe that was something I copied from the draw function. Eliminating the `math.floor` is definitely smarter, I'll be sure to fix that.

Aire-One commented:
Since the PR is still a work in progress, I'm converting it to a draft.

Elv13 commented:
>  Are the code examples supposed to be in the widget file?

`--@DOC_wibox_widget_slider_handle_widget_EXAMPLE@` looks for a file called `tests/examples/wibox/widget/slider/handle_widget.lua` (it just converts the underscore to slashes). You can copy paste another file from that directory and modify the content. The build system will take care of prettyfying the code, generate the image and insert it in the doc. You need to have `ldoc` installed for this to work. It will also print some warnings if it doesn't like how the doc is written. If it doesn't then you are likely doing it right (or I need to add more warnings to my linting bot).

thomascft commented:
Will these also add object properties?

Elv13 commented:
> Will these also add object properties?

They don't add anything, they are just code examples used to generate images and (optionally) code for the doc. This allows us to unit test our examples and have global templates to maintain the images. If you clean your build directory, copy paste some existing files from the same directory and rebuild awesome, it should just work. Then you can modify the code to better suit the property you are demonstrating.

thomascft commented:
I'm having and unreasonably difficult time getting `handle_widget` and `bar_widget` to show up in the docs. 
```lua
-- The slider handle widget.
--
--@DOC_wibox_widget_slider_handle_widget_EXAMPLE@
--
-- @property handle_widget
-- @tparam widget|nil handle_widget
-- @propemits true false
-- @propbeautiful
-- @see wibox.widget
```
The file I'm opening to check is `build/doc/widgets/wibox.widget.slider.html`
What am I doing wrong?

thomascft commented:
I'm not 100% sure on the method documentation for `set_handle_margins` or `set_bar_margins`, I can only get the table parameter to render in the docs. If y'all could take a look at that, I think it's mostly ready.

thomascft commented:
Hold up, I forgot testing.

Elv13 commented:
The image is `./raw_images/AUTOGEN_wibox_widget_slider_handle_widget.svg`. It was getting generated, but empty because nothing called the `gen()` function. I copy pasted the code back and it does things:

```patch
diff --git a/tests/examples/wibox/widget/slider/bar_widget.lua b/tests/examples/wibox/widget/slider/bar_widget.lua
index 96415c17f..d950b4e8b 100644
--- a/tests/examples/wibox/widget/slider/bar_widget.lua
+++ b/tests/examples/wibox/widget/slider/bar_widget.lua
@@ -20,4 +20,19 @@ local function gen(val)
     }
 end
 
+local l = wibox.layout {
+    gen(0), gen(3), gen(6), gen({
+                                top    = 12,
+                                bottom = 12,
+                                left   = 12,
+                                right  = 12,
+                            }),
+    forced_height = 30,
+    forced_width  = 400,
+    spacing       = 5,
+    layout        = wibox.layout.flex.horizontal
+}
+
+parent:add(l)
+
 --DOC_HIDE vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
diff --git a/tests/examples/wibox/widget/slider/handle_widget.lua b/tests/examples/wibox/widget/slider/handle_widget.lua
index 29928c1b7..05d0c7567 100644
--- a/tests/examples/wibox/widget/slider/handle_widget.lua
+++ b/tests/examples/wibox/widget/slider/handle_widget.lua
@@ -20,4 +20,19 @@ local function gen(val)
     }
 end
 
+local l = wibox.layout {
+    gen(0), gen(3), gen(6), gen({
+                                top    = 12,
+                                bottom = 12,
+                                left   = 12,
+                                right  = 12,
+                            }),
+    forced_height = 30,
+    forced_width  = 400,
+    spacing       = 5,
+    layout        = wibox.layout.flex.horizontal
+}
+
+parent:add(l)
+
 --DOC_HIDE vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
```

However, "does things" here means this:

> 2022-10-02 22:25:00 E: awesome: Error during a protected call: /home/lepagee/dev/awesome/lib/wibox/widget/slider.lua:643: attempt to perform arithmetic on a nil value (field 'left')
stack traceback:
        /home/lepagee/dev/awesome/lib/wibox/widget/slider.lua:643: in function 'wibox.widget.slider.layout'
        [C]: in function 'xpcall'
        /home/lepagee/dev/awesome/lib/gears/protected_call.lua:41: in function </home/lepagee/dev/awesome/lib/gears/protected_call.lua:40>
        (...tail calls...)
        /home/lepagee/dev/awesome/lib/gears/cache.lua:42: in function 'gears.cache.get'
        (...tail calls...)
        /home/lepagee/dev/awesome/lib/wibox/hierarchy.lua:133: in upvalue 'hierarchy_update'
        /home/lepagee/dev/awesome/lib/wibox/hierarchy.lua:141: in upvalue 'hierarchy_update'
        /home/lepagee/dev/awesome/lib/wibox/hierarchy.lua:141: in upvalue 'hierarchy_update'
        /home/lepagee/dev/awesome/lib/wibox/hierarchy.lua:230: in function 'wibox.hierarchy.update'
        /home/lepagee/dev/awesome/lib/wibox/hierarchy.lua:215: in function 'wibox.hierarchy.new'
        /home/lepagee/dev/awesome/lib/gears/surface.lua:229: in upvalue 'run_in_hierarchy'
        /home/lepagee/dev/awesome/lib/gears/surface.lua:257: in function 'gears.surface.widget_to_svg'
        /home/lepagee/dev/awesome/tests/examples/wibox/template.lua:62: in main chunk
        [C]: in ?

This is what tests are for ;)

Elv13 commented on lib/wibox/widget/slider.lua:
```diff
@@ -307,7 +323,47 @@ for prop in pairs(properties) do
     end
 end
 
+--- Set the handle widget
```
Those `@method` sections are not required. The `@property` is enough. Otherwise, it will document how to set the widget twice. We only document some legacy setters which had more than 1 parameters. Everything else is documented using the property mechanism.

thomascft commented:
Shouldn't that be 
```lua
gen ({ margins = {
    left = ...
    }
})

thomascft commented:
> The image is `./raw_images/AUTOGEN_wibox_widget_slider_handle_widget.svg`. It was getting generated, but empty because nothing called the `gen()` function. I copy pasted the code back and it does things:
> 
> This is what tests are for ;)

I think the margins aren't being created by their setter functions. in the setter function I set the fallback of 0.

Elv13 commented:
> I think the margins aren't being created by their setter functions. in the setter function I set the fallback of 0.

Yeah, it will need an `for _, prop in ipairs {"handle_margins", "bar_margins"} do if not args[prop] then ret[prop] = 0 end end`. I see you mentioned this earlier, I apologize for not connecting the dots.

thomascft commented:
that'll set it to  0, but it needs to build the table aswell.

thomascft commented:
Where should I put that? 

> > I think the margins aren't being created by their setter functions. in the setter function I set the fallback of 0.
> 
> Yeah, it will need an `for _, prop in ipairs {"handle_margins", "bar_margins"} do if not args[prop] then ret[prop] = 0 end end`. I see you mentioned this earlier, I apologize for not connecting the dots.

The args is always nil

Elv13 commented:
> The args is always nil

I fixed that in the diff below by adding `args = args or {}`

```patch
diff --git a/lib/wibox/widget/slider.lua b/lib/wibox/widget/slider.lua
index 344f91e20..7becf2e9e 100644
--- a/lib/wibox/widget/slider.lua
+++ b/lib/wibox/widget/slider.lua
@@ -40,7 +40,7 @@ local slider = {mt={}}
 --@DOC_wibox_widget_slider_handle_widget_EXAMPLE@
 --
 -- @property handle_widget
--- @tparam[opt=nil] widget|nil handle_widget The handle widget 
+-- @tparam[opt=nil] widget|nil handle_widget The handle widget
 
 --- The slider handle color.
 --
@@ -323,14 +323,6 @@ for prop in pairs(properties) do
     end
 end
 
---- Set the handle widget
---
--- @method set_handle_widget
--- @tparam[opt=nil] widget|nil bar_widget Set the handle widget
--- @noreturn
-
-
-
 function slider:set_handle_widget(value)
     local changed = self._private.handle_widget ~= value
         self._private.handle_widget = value
@@ -340,14 +332,6 @@ function slider:set_handle_widget(value)
     end
 end
 
---- Set the bar widget
---
--- @method set_bar_widget
--- @tparam[opt=nil] widget|nil bar_widget Set the widget that spans the active bar segment
--- @noreturn
-
-
-
 function slider:set_bar_widget(value)
     local changed = self._private.bar_widget ~= value
         self._private.bar_widget = value
@@ -357,13 +341,6 @@ function slider:set_bar_widget(value)
     end
 end
 
--- Add some validation to set_value
-
---- Set the slider's value
--- @tparam number number Set value to number
--- @method set_value
--- @noreturn
-
 function slider:set_value(value)
     value = math.min(value, self:get_maximum())
     value = math.max(value, self:get_minimum())
@@ -388,22 +365,9 @@ local function get_extremums(self)
     return min, max, interval
 end
 
--- TODO add single number param to margin setters
-
---- Set the slider's margins
---
--- @tparam[opt={}] table|number|nil handle_margins
--- @tparam[opt=0] number handle_margins.left
--- @tparam[opt=0] number handle_margins.right
--- @tparam[opt=0] number handle_margins.top
--- @tparam[opt=0] number handle_margins.bottom
--- @propemits true false
--- @propbeautiful
--- @method set_handle_margins
--- @noreturn
-
 function slider:set_handle_margins(value)
-    local value = value or 0
+    value = value or 0
+
     if type(value) == "number" then
         value = {
             top    = value,
@@ -419,19 +383,9 @@ function slider:set_handle_margins(value)
     self:emit_signal("widget::layout_changed")
 end
 
---- Set the bar's margins
---
--- @tparam[opt={}] table|number|nil bar_margins
--- @tparam[opt=0] number bar_margins.left
--- @tparam[opt=0] number bar_margins.right
--- @tparam[opt=0] number bar_margins.top
--- @tparam[opt=0] number bar_margins.bottom
--- @propemits true false
--- @propbeautiful
--- @method set_bar_margins
--- @noreturn
-
 function slider:set_bar_margins(value)
+    value = value or 0
+
     if type(value) == "number" then
         value = {
             top    = value,
@@ -625,13 +579,15 @@ function slider:layout(context, width, height)
 
     local value = self._private.value or self._private.min or 0
     local min, _, interval = get_extremums(self)
-    local bar_height, bar_width = self._private.bar_height or height, width
+    local bar_height = self._private.bar_height or height
     local handle_height, handle_width = height, self._private.handle_width
         or beautiful.slider_handle_width
         or width
 
     if bar_widget then
-        bar_width = (((value - min) / interval) * width + handle_width / 2 - ((value - min) / interval ) * handle_width)
+        local bar_width = (
+            ((value - min) / interval) * width + handle_width / 2
+                - ((value - min) / interval ) * handle_width)
         local w, h = base.fit_widget(self, context, bar_widget, bar_width, bar_height)
         local margins = self._private.bar_margins
             or beautiful.slider_bar_margins
@@ -639,9 +595,7 @@ function slider:layout(context, width, height)
         local x_offset, y_offset = 0, 0
 
         if margins then
-                x_offset, y_offset = margins.left, margins.top
-                bar_width       = bar_width -
-                    (margins.left) - (margins.right)
+            x_offset, y_offset = margins.left, margins.top
         end
         local x, y = 0 + x_offset, 0 + y_offset
         table.insert(result, base.place_widget_at(bar_widget, x, y, w, h))
@@ -725,6 +679,8 @@ end
 -- @tparam[opt] number args.minimum The slider minimum value.
 -- @tparam[opt] number args.maximum The slider maximum value.
 local function new(args)
+    args = args or {}
+
     local ret = base.make_widget(nil, nil, {
         enable_properties = true,
     })
@@ -735,6 +691,13 @@ local function new(args)
 
     ret:connect_signal("button::press", mouse_press)
 
+    -- Initialize the tables.
+    for _, prop in ipairs {"handle_margins", "bar_margins"} do
+        if not args[prop] then
+           ret[prop] = 0
+        end
+    end
+
     return ret
 end
 
```

With that, the `handle_widget` image renders.

I also, there's a little helpful command called `luacheck` which you can run before sending code to GitHub, it will catch a few warnings which would have failed the CI. There was just a couple, I fixed them in the diff above. Do you want me to push this patch to the PR?

thomascft commented:
Sure

thomascft commented:
> > The args is always nil
> 
> I fixed that in the diff below by adding `args = args or {}`
> 
> ```diff
> diff --git a/lib/wibox/widget/slider.lua b/lib/wibox/widget/slider.lua
> index 344f91e20..7becf2e9e 100644
> --- a/lib/wibox/widget/slider.lua
> +++ b/lib/wibox/widget/slider.lua
> @@ -40,7 +40,7 @@ local slider = {mt={}}
>  --@DOC_wibox_widget_slider_handle_widget_EXAMPLE@
>  --
>  -- @property handle_widget
> --- @tparam[opt=nil] widget|nil handle_widget The handle widget 
> +-- @tparam[opt=nil] widget|nil handle_widget The handle widget
>  
>  --- The slider handle color.
>  --
> @@ -323,14 +323,6 @@ for prop in pairs(properties) do
>      end
>  end
>  
> ---- Set the handle widget
> ---
> --- @method set_handle_widget
> --- @tparam[opt=nil] widget|nil bar_widget Set the handle widget
> --- @noreturn
> -
> -
> -
>  function slider:set_handle_widget(value)
>      local changed = self._private.handle_widget ~= value
>          self._private.handle_widget = value
> @@ -340,14 +332,6 @@ function slider:set_handle_widget(value)
>      end
>  end
>  
> ---- Set the bar widget
> ---
> --- @method set_bar_widget
> --- @tparam[opt=nil] widget|nil bar_widget Set the widget that spans the active bar segment
> --- @noreturn
> -
> -
> -
>  function slider:set_bar_widget(value)
>      local changed = self._private.bar_widget ~= value
>          self._private.bar_widget = value
> @@ -357,13 +341,6 @@ function slider:set_bar_widget(value)
>      end
>  end
>  
> --- Add some validation to set_value
> -
> ---- Set the slider's value
> --- @tparam number number Set value to number
> --- @method set_value
> --- @noreturn
> -
>  function slider:set_value(value)
>      value = math.min(value, self:get_maximum())
>      value = math.max(value, self:get_minimum())
> @@ -388,22 +365,9 @@ local function get_extremums(self)
>      return min, max, interval
>  end
>  
> --- TODO add single number param to margin setters
> -
> ---- Set the slider's margins
> ---
> --- @tparam[opt={}] table|number|nil handle_margins
> --- @tparam[opt=0] number handle_margins.left
> --- @tparam[opt=0] number handle_margins.right
> --- @tparam[opt=0] number handle_margins.top
> --- @tparam[opt=0] number handle_margins.bottom
> --- @propemits true false
> --- @propbeautiful
> --- @method set_handle_margins
> --- @noreturn
> -
>  function slider:set_handle_margins(value)
> -    local value = value or 0
> +    value = value or 0
> +
>      if type(value) == "number" then
>          value = {
>              top    = value,
> @@ -419,19 +383,9 @@ function slider:set_handle_margins(value)
>      self:emit_signal("widget::layout_changed")
>  end
>  
> ---- Set the bar's margins
> ---
> --- @tparam[opt={}] table|number|nil bar_margins
> --- @tparam[opt=0] number bar_margins.left
> --- @tparam[opt=0] number bar_margins.right
> --- @tparam[opt=0] number bar_margins.top
> --- @tparam[opt=0] number bar_margins.bottom
> --- @propemits true false
> --- @propbeautiful
> --- @method set_bar_margins
> --- @noreturn
> -
>  function slider:set_bar_margins(value)
> +    value = value or 0
> +
>      if type(value) == "number" then
>          value = {
>              top    = value,
> @@ -625,13 +579,15 @@ function slider:layout(context, width, height)
>  
>      local value = self._private.value or self._private.min or 0
>      local min, _, interval = get_extremums(self)
> -    local bar_height, bar_width = self._private.bar_height or height, width
> +    local bar_height = self._private.bar_height or height
>      local handle_height, handle_width = height, self._private.handle_width
>          or beautiful.slider_handle_width
>          or width
>  
>      if bar_widget then
> -        bar_width = (((value - min) / interval) * width + handle_width / 2 - ((value - min) / interval ) * handle_width)
> +        local bar_width = (
> +            ((value - min) / interval) * width + handle_width / 2
> +                - ((value - min) / interval ) * handle_width)
>          local w, h = base.fit_widget(self, context, bar_widget, bar_width, bar_height)
>          local margins = self._private.bar_margins
>              or beautiful.slider_bar_margins
> @@ -639,9 +595,7 @@ function slider:layout(context, width, height)
>          local x_offset, y_offset = 0, 0
>  
>          if margins then
> -                x_offset, y_offset = margins.left, margins.top
> -                bar_width       = bar_width -
> -                    (margins.left) - (margins.right)
> +            x_offset, y_offset = margins.left, margins.top
>          end
>          local x, y = 0 + x_offset, 0 + y_offset
>          table.insert(result, base.place_widget_at(bar_widget, x, y, w, h))
> @@ -725,6 +679,8 @@ end
>  -- @tparam[opt] number args.minimum The slider minimum value.
>  -- @tparam[opt] number args.maximum The slider maximum value.
>  local function new(args)
> +    args = args or {}
> +
>      local ret = base.make_widget(nil, nil, {
>          enable_properties = true,
>      })
> @@ -735,6 +691,13 @@ local function new(args)
>  
>      ret:connect_signal("button::press", mouse_press)
>  
> +    -- Initialize the tables.
> +    for _, prop in ipairs {"handle_margins", "bar_margins"} do
> +        if not args[prop] then
> +           ret[prop] = 0
> +        end
> +    end
> +
>      return ret
>  end
>  
> ```
> 
> With that, the `handle_widget` image renders.
> 
> I also, there's a little helpful command called `luacheck` which you can run before sending code to GitHub, it will catch a few warnings which would have failed the CI. There was just a couple, I fixed them in the diff above. Do you want me to push this patch to the PR?

Is there anything else you'd like me to get working before the merge?

Elv13 commented:
No, I need to circle back to this. But we have another PR in progress for a generc template system. It would be nice to have this merged first so we can use it in the slider. This would allow to have the bar widget to display the slider value inline (optionally) and that kind of stuff.

thomascft commented:
Good to know
